### ISSUE TITLE
[rlgl] Array Texture support

### ISSUE THREAD

**Issue by julhe** [raysan5/raylib#3402]
## Issue description

For my project I need [array texture](https://www.khronos.org/opengl/wiki/Array_Texture) support. 

Array textures simplfy managing multiple textures of the same size and format. Instead of switching the texture each drawcall, a single index can be used and thus objects with different textures can be batched. You can even use them to emulate 3D texture without much overhead. 

This feature is core since OpenGL 3.0 and OpenGL ES 3.0, which both are supported by Raylib. However they are not exposed by Raylib.

## Proposed changes
I could think of two ways to add this to raylib:

* ### Add a new ``Texture2DArray`` type:
    ```C
    typedef struct Texture2DArray {
        unsigned int id;        // OpenGL texture id
        int width;              // Texture base width
        int height;             // Texture base height
        int slices;             // Amount of slices
        int mipmaps;            // Mipmap levels, 1 by default
        int format;             // Data format (PixelFormat type)
    } Texture2DArray;
    ``` 

    👍Pros:
    * Doesn't touch ```Texture```. 
    * Type safety at compile time.
    * Can be hidden away in rlgl to keep the core API clean.
    * Easy to validate platform compatibility at compile time.

    👎Cons:
    * Needs additional API which increases rlgl complexity.
  
* ### Alternative: Turn ``Texture`` into an uber struct for all texture needs:
    ```C
    typedef struct Texture {
        unsigned int id;        // OpenGL texture id
        int width;              // Texture base width
        int height;             // Texture base height
        int slices;             // NEW: Amount of slices, 0 by default
        int mipmaps;            // Mipmap levels, 1 by default
        int format;             // Data format (PixelFormat type)
    } Texture;
    ``` 
    By setting ``slices`` to > 1 indicates that this is an array texture.

    👍Pros:
    * No additional API. Existing texture API would simply be extended for support.
    
    👎Cons:
    * Increases complexity of ```Texture``` which is frequently used.
    * No type safety. Mismatching types (texture vs. array texture) would only be caught at runtime.
    * Might confuse beginners.

**Comment by raysan5** [comment]
@julhe I prefer to avoid texture arrays. I think not that many users would benefit of this addition and it adds an unneeded extra level of complexity to raylib.

**Comment by Clemens-L** [comment]
@raysan5 I was just researching array texture support in raylib and stumbled upon this issue.
In my opinion, support for array textures could actually benefit many users, as it is the preferred alternative to using texture atlases, avoiding issues such as bleeding during mipmap generation.

I understand the desire to limit the complexity of raylib, however, this also limits the what can be achieved with it and/or greatly increases the complexity of applications developed with it (for example, having to deal with mipmapping texture atlases correctly).

I really enjoy working with raylib as it usually speeds up my work a lot - in some situations, such as this one, it can be frustrating though. 

As texture arrays are nowadays a rather standard technique, I would argue that it could actually be confusing for users not to be able to use it, and having to learn about workarounds / other techniques to accomplish the same goal. 

A reconsideration of this feature on your part would be greatly appreciated, but I understand if you choose to limit raylib to more basic features.

**Comment by raysan5** [comment]
I think the use of texture arrays is limited to a few specific use-cases, most users don't probably need it and users requiring that advance feature are probably users able to implement it directly with OpenGL for their specific use cases.

**Comment by casperbear** [comment]
I implemented texture arrays in a fork for my own needs and just wanted to let everyone interested know that there is nothing tricky about them. You just load them with glTexImage3D/glTexSubImage3D instead of glTexImage2D and then to bind / unbind / delete / set wrapping / generate mipmaps / set trilinear/anisotropic filtering you use all the same calls, just pass GL_TEXTURE_2D_ARRAY instead of GL_TEXTURE_2D. Copy-paste-AI-copy-paste. The only moment I had to think is to decide that to pass slice number to the shaders texcoords2.x should be used (in my case). Texture arrays require OpenGL 3.3 though.
